File: bots/beacon/bestrn/util/directions.py
"""Translation of `bots/intgrah/v54.7.9/util/directions.py`."""

from __future__ import annotations

from typing import Final

from cambc import Direction

DIR8: Final[list[Direction]] = [
    Direction.NORTH,
    Direction.NORTHEAST,
    Direction.EAST,
    Direction.SOUTHEAST,
    Direction.SOUTH,
    Direction.SOUTHWEST,
    Direction.WEST,
    Direction.NORTHWEST,
]
"""
`N, NE, E, SE, S, SW, W, NW` — the eight non-CENTRE directions, in
`EntityType` enum order (matches Python `[d for d in Direction if d != CENTRE]`).
"""
DIR4: Final[list[Direction]] = [
    Direction.NORTH,
    Direction.EAST,
    Direction.SOUTH,
    Direction.WEST,
]
"""Cardinal directions only: `N, E, S, W`."""


def is_cardinal(d):
    """
    True for the four cardinal directions (N/E/S/W). Mirrors the Rust
    `Direction::is_cardinal()` helper from `libre-engine`, which has no
    Python counterpart on `cambc.Direction`.
    """
    return d in (Direction.NORTH, Direction.EAST, Direction.SOUTH, Direction.WEST)


DIR8_DELTA: Final[list[tuple[int, int]]] = [
    (0, -1),
    (1, -1),
    (1, 0),
    (1, 1),
    (0, 1),
    (-1, 1),
    (-1, 0),
    (-1, -1),
]
"""`(dx, dy)` offsets for `DIR8`, in the same order."""


def delta_to_dir(dx, dy):
    """
    Convert a magnitude-1 or sqrt-2 vector to its `Direction`. Returns `None`
    for any other delta (including `(0, 0)` and longer vectors).
    """
    match (dx, dy):
        case (0, -1):
            return Direction.NORTH
        case (1, -1):
            return Direction.NORTHEAST
        case (1, 0):
            return Direction.EAST
        case (1, 1):
            return Direction.SOUTHEAST
        case (0, 1):
            return Direction.SOUTH
        case (-1, 1):
            return Direction.SOUTHWEST
        case (-1, 0):
            return Direction.WEST
        case (-1, -1):
            return Direction.NORTHWEST
        case _:
            return None


def get_direction_object(from_pos, to_pos):
    """Direction from `from_pos` to `to_pos`, or `None` if not adjacent in king-move."""
    return delta_to_dir(to_pos.x - from_pos.x, to_pos.y - from_pos.y)

File: bots/beacon/v670/src/util/directions.rs
//! Translation of `bots/intgrah/v54.7.9/util/directions.py`.

use cambc::{Direction, Position};

/// `N, NE, E, SE, S, SW, W, NW` — the eight non-CENTRE directions, in
/// `EntityType` enum order (matches Python `[d for d in Direction if d != CENTRE]`).
pub const DIR8: [Direction; 8] = [
    Direction::North,
    Direction::Northeast,
    Direction::East,
    Direction::Southeast,
    Direction::South,
    Direction::Southwest,
    Direction::West,
    Direction::Northwest,
];

/// Cardinal directions only: `N, E, S, W`.
pub const DIR4: [Direction; 4] = [
    Direction::North,
    Direction::East,
    Direction::South,
    Direction::West,
];

/// True for the four cardinal directions (N/E/S/W). Mirrors the Rust
/// `Direction::is_cardinal()` helper from `cambc-libre-engine`, which has no
/// Python counterpart on `cambc.Direction`.
#[must_use]
pub const fn is_cardinal(d: Direction) -> bool {
    matches!(
        d,
        Direction::North | Direction::East | Direction::South | Direction::West
    )
}

/// Rotate a direction 45° clockwise. `Centre` is a no-op.
#[must_use]
pub const fn rotate_right(d: Direction) -> Direction {
    match d {
        Direction::North => Direction::Northeast,
        Direction::Northeast => Direction::East,
        Direction::East => Direction::Southeast,
        Direction::Southeast => Direction::South,
        Direction::South => Direction::Southwest,
        Direction::Southwest => Direction::West,
        Direction::West => Direction::Northwest,
        Direction::Northwest => Direction::North,
        Direction::Centre => Direction::Centre,
    }
}

/// `(dx, dy)` offsets for `DIR8`, in the same order.
pub const DIR8_DELTA: [(i32, i32); 8] = [
    (0, -1),  // North
    (1, -1),  // Northeast
    (1, 0),   // East
    (1, 1),   // Southeast
    (0, 1),   // South
    (-1, 1),  // Southwest
    (-1, 0),  // West
    (-1, -1), // Northwest
];

/// `(dx, dy)` offsets for `DIR4` (cardinals only), in the same order.
pub const DIR4_DELTA: [(i32, i32); 4] = [
    (0, -1), // North
    (1, 0),  // East
    (0, 1),  // South
    (-1, 0), // West
];

/// Convert a magnitude-1 or sqrt-2 vector to its `Direction`. Returns `None`
/// for any other delta (including `(0, 0)` and longer vectors).
#[must_use]
pub const fn delta_to_dir(dx: i32, dy: i32) -> Option<Direction> {
    match (dx, dy) {
        (0, -1) => Some(Direction::North),
        (1, -1) => Some(Direction::Northeast),
        (1, 0) => Some(Direction::East),
        (1, 1) => Some(Direction::Southeast),
        (0, 1) => Some(Direction::South),
        (-1, 1) => Some(Direction::Southwest),
        (-1, 0) => Some(Direction::West),
        (-1, -1) => Some(Direction::Northwest),
        _ => None,
    }
}

/// Direction from `from_pos` to `to_pos`, or `None` if not adjacent in king-move.
#[must_use]
pub const fn get_direction_object(from_pos: Position, to_pos: Position) -> Option<Direction> {
    delta_to_dir(to_pos.x - from_pos.x, to_pos.y - from_pos.y)
}
